wgpu-hal/src/dx12/mod.rs

```diff
@@ -1275,6 +1290,128 @@ impl crate::Surface for Surface {
+                                                "IDCompositionDevice::CreateTargetForHwnd",
+                                            )
+                                        })?
+                                };
+
+                                let visual = {
+                                    profiling::scope!("IDCompositionDevice::CreateVisual");
+                                    unsafe { dcomp_device.CreateVisual() }.map_err(|err| {
+                                        log::error!(
+                                            "IDCompositionDevice::CreateVisual failed: {err}"
+                                        );
+                                        crate::SurfaceError::Other(
+                                            "IDCompositionDevice::CreateVisual",
+                                        )
+                                    })?
+                                };
+
+                                {
+                                    profiling::scope!("IDCompositionTarget::SetRoot");
+                                    unsafe { target.SetRoot(&visual) }.map_err(|err| {
+                                        log::error!("IDCompositionTarget::SetRoot failed: {err}");
+                                        crate::SurfaceError::Other("IDCompositionTarget::SetRoot")
+                                    })?;
+                                }
+
+                                dcomp_state.insert(DCompState {
+                                    visual,
+                                    device: dcomp_device,
+                                    _target: target,
+                                })
+                            }
+                        };
+                        // Set the new swap chain as the content for the backing visual
+                        // and commit the changes to the composition visual tree.
+                        {
+                            profiling::scope!("IDCompositionVisual::SetContent");
+                            unsafe { dcomp_state.visual.SetContent(&swap_chain1) }.map_err(
+                                |err| {
+                                    log::error!("IDCompositionVisual::SetContent failed: {err}");
+                                    crate::SurfaceError::Other("IDCompositionVisual::SetContent")
+                                },
+                            )?;
+                        }
+
+                        // Commit the changes to the composition device.
+                        {
+                            profiling::scope!("IDCompositionDevice::Commit");
+                            unsafe { dcomp_state.device.Commit() }.map_err(|err| {
+                                log::error!("IDCompositionDevice::Commit failed: {err}");
+                                crate::SurfaceError::Other("IDCompositionDevice::Commit")
+                            })?;
+                        }
```

Comment:
Could we contain all of this dcomp code into a `dcomp` module, so that this is cleaner.